library/psa_crypto.c

```diff
@@ -3806,7 +3802,7 @@ static psa_status_t psa_aead_setup( psa_aead_operation_t *operation,
     if( status != PSA_SUCCESS )
         goto exit;
 
-    if( ( status = psa_validate_tag_length( operation, alg ) ) != PSA_SUCCESS )
+    if( ( status = psa_validate_tag_length( alg ) ) != PSA_SUCCESS )
```

Comment:
I think that this should be done before to call the driver set up functions. That way we validate all we can validate here in the core before to call the driver. This should probably be integrated into `psa_aead_check_algorithm()`.

Furthermore, the algorithm tag length being validated by the core for CCM, GCM and CHACHA20_POLY1305 further validation in the drivers and builtin implementation is not necessary. Thus the following check in `psa_crypto_aead.c:psa_aead_setup()` can probably be removed:
```C
   if( PSA_AEAD_TAG_LENGTH( attributes->core.type,                             
                             key_bits, alg )                                    
        > full_tag_length )                                                     
        return( PSA_ERROR_INVALID_ARGUMENT );
```
And probably that at the end of `psa_crypto_aead.c:psa_aead_setup()`
```C
    operation->tag_length = PSA_AEAD_TAG_LENGTH( operation->key_type,           
                                                 key_bits,                      
                                                 alg );
```
should be replaced by just:
```C
    operation->tag_length = PSA_ALG_AEAD_GET_TAG_LENGTH(alg);
```

Reply:
This will require to change the test results:
```
PSA Multipart AEAD verify: ChaCha20 - Poly1305, invalid tag length  FAILED
  status == expected_setup_status
  at line 5566, suites/test_suite_psa_crypto.function
  lhs = 0xffffffffffffff79 = -135
  rhs = 0xffffffffffffff7a = -134
PSA Multipart AEAD verify: ChaCha20 - Poly1305, invalid tag length  FAILED
  status == expected_setup_status
  at line 5566, suites/test_suite_psa_crypto.function
  lhs = 0xffffffffffffff79 = -135
  rhs = 0xffffffffffffff7a = -134
PSA Multipart AEAD verify: ChaCha20 - Poly1305, invalid tag length  FAILED
  status == expected_setup_status
  at line 5566, suites/test_suite_psa_crypto.function
  lhs = 0xffffffffffffff79 = -135
  rhs = 0xffffffffffffff7a = -134

PSA AEAD setup: invalid algorithm (ChaCha20 - Poly1305 with short   FAILED
  status == expected_status
  at line 5633, suites/test_suite_psa_crypto.function
  lhs = 0xffffffffffffff79 = -135
  rhs = 0xffffffffffffff7a = -134

PSA AEAD setup: ChaCha20-Poly1305, invalid tag length 0 ........... FAILED
  status == expected_status
  at line 5633, suites/test_suite_psa_crypto.function
  lhs = 0xffffffffffffff79 = -135
  rhs = 0xffffffffffffff7a = -134
PSA AEAD setup: ChaCha20-Poly1305, invalid tag length 15 .......... FAILED
  status == expected_status
  at line 5633, suites/test_suite_psa_crypto.function
  lhs = 0xffffffffffffff79 = -135
  rhs = 0xffffffffffffff7a = -134
PSA AEAD setup: ChaCha20-Poly1305, invalid tag length 17 .......... FAILED
  S== expected_status
  at line 5633, suites/test_suite_psa_crypto.function
  lhs = 0xffffffffffffff79 = -135
  rhs = 0xffffffffffffff7a = -134

```
In above test cases `status` is now invalid argument.